File: 8/4.py
def main(n):
    def dfs(j=1, result=[]):
        if j > n:
            return
        for i in range(j, 4):
            result.append(i)
            dfs(i + 1, result)
            print(result)
            result.pop()

    dfs()


print(main(3))

File: 8/4.js
// 시간 복잡도: O(3^n)
// 공간 복잡도: O(n)
function main(n) {
  function dfs(j = 1, result = []) {
    if (j > n) {
      return;
    }

    for (let i = j; i <= 3; i++) {
      result.push(i);
      dfs(i + 1, result);
      console.log(result);
      result.pop();
    }
  }

  dfs();
}

console.log(main(3));
